src/main/network/dns.rs

```diff
@@ -0,0 +1,170 @@
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::Write;
+use std::net::Ipv4Addr;
+use std::os::fd::AsRawFd;
+use std::path::PathBuf;
+use std::sync::Arc;
+
+use rustix::fs::MemfdFlags;
+use shadow_shim_helper_rs::HostId;
+
+#[derive(Debug)]
+struct Database {
+    name_index: HashMap<String, Arc<Record>>,
```

Comment:
This wan't meant to impose a restriction. Although there are other restrictions listed in `hostname(7)`, those would probably be better checked on yaml validation if we wanted that rather than here in a submodule.

Would you recommend `OsString` here instead?

Reply:
FWIW, `HostInfo.name` is already a `String` and that is what we are passing into the `register()` function.

Reply:
I guess this depends on what we want the dns module to support. Do we only want to support standard hostnames ([a-zA-Z_-.] or whatever), or any valid domain name?

If it's any valid domain name, then I don't know what we should use. It sounds like the ccTLD can decide that their domain names can use whatever they want, so we'd probably want to use `Vec<u8>` in that case. Or since we're linux-only we could use `OsString` since on linux it's just a byte string.

But I don't really think it's worth worrying about any of that, and we should just use `String` instead.

How about adding a comment like: "We can use `String` here because [`crate::core::configuration::HostName`] limits the configured host names to a subset of ascii, which are always valid utf-8"?